crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -1,232 +1,1340 @@
+    let mut current = HashSet::with_capacity_and_hasher(graph.len(), Default::default());
+    let mut unvisited = HashSet::with_capacity_and_hasher(graph.len(), Default::default());
+    unvisited.extend(graph.keys().cloned());
+    while let Some(node) = unvisited.iter().next().cloned() {
+        if check_if_cycles_and_visit(&node, graph, &mut sorted, &mut unvisited, &mut current) {
+            return Err(DependencyGraphError::GraphCycles(
+                current
+                    .iter()
+                    .map(|index| systems[*index].display_name())
+                    .collect::<Vec<_>>(),
+            ));
+        }
+    }
+    Ok(sorted)
 }
 
-impl<S: System<In = (), Out = ()>> From<S> for SystemStage {
-    fn from(system: S) -> Self {
-        SystemStage::single(system)
+/// Returns vector containing all pairs of indices of systems with ambiguous execution order.
+/// Systems must be topologically sorted beforehand.
+fn find_ambiguities(systems: &[impl SystemContainer]) -> Vec<(usize, usize)> {
+    let mut all_dependencies = Vec::<FixedBitSet>::with_capacity(systems.len());
+    let mut all_dependants = Vec::<FixedBitSet>::with_capacity(systems.len());
+    for (index, container) in systems.iter().enumerate() {
+        let mut dependencies = FixedBitSet::with_capacity(systems.len());
+        for &dependency in container.dependencies() {
+            dependencies.union_with(&all_dependencies[dependency]);
+            dependencies.insert(dependency);
+            all_dependants[dependency].insert(index);
+        }
+        all_dependants.push(FixedBitSet::with_capacity(systems.len()));
+        all_dependencies.push(dependencies);
     }
+    for index in (0..systems.len()).rev() {
+        let mut dependants = FixedBitSet::with_capacity(systems.len());
+        for dependant in all_dependants[index].ones() {
+            dependants.union_with(&all_dependants[dependant]);
+            dependants.insert(dependant);
+        }
+        all_dependants[index] = dependants;
+    }
+    let mut all_relations = all_dependencies
+        .drain(..)
+        .zip(all_dependants.drain(..))
+        .enumerate()
+        .map(|(index, (dependencies, dependants))| {
+            let mut relations = FixedBitSet::with_capacity(systems.len());
+            relations.union_with(&dependencies);
+            relations.union_with(&dependants);
+            relations.insert(index);
+            relations
+        })
+        .collect::<Vec<FixedBitSet>>();
+    let mut ambiguities = Vec::new();
+    let full_bitset: FixedBitSet = (0..systems.len()).collect();
+    for (index_a, relations) in all_relations.drain(..).enumerate() {
+        // TODO: prove that `.take(index_a)` would be correct here, and uncomment it if so.
+        for index_b in full_bitset.difference(&relations)
+        /*.take(index_a)*/
+        {
+            if !systems[index_a].is_compatible(&systems[index_b])
+                && !ambiguities.contains(&(index_b, index_a))
```

Comment:
we can cut down on the runtime complexity of this using a bitset:

```rust
let mut processed = FixedBitSet::with_capacity(systems.len());
for (index_a, relations) in all_relations.drain(..).enumerate() {
    // TODO: prove that `.take(index_a)` would be correct here, and uncomment it if so.
    for index_b in full_bitset.difference(&relations)
    /*.take(index_a)*/
    {
        if !processed.contains(index_b) && !systems[index_a].is_compatible(&systems[index_b])
        {
            ambiguities.push((index_a, index_b));
        }
    }
    processed.insert(index_a);
}
```

Ambiguities are always "double reported" without deduping logic, so if we have already index_b, we know it has already been checked for compatibility and added to the ambiguity list (when relevant).

Tests pass with this change. And using Bevy as a test case, the total number of ambiguities reported for each stage is the same across both implementations.

Reply:
Done in 69a55aa1f6023e3dae6679d42e6d1db0d7ca7e02